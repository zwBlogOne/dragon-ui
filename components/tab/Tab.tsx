import React, { Component } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';

class Tab extends Component<PropsType, any> {
  static Group;

  static defaultProps = {
    prefixCls: 'ui-tab',
  };

  constructor(props) {
    super(props);
    this.state = {
      selected: props.selected || props.defaultSelected,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('selected' in nextProps) {
      this.setState({
        selected: !!nextProps.selected,
      });
    }
  }

  render() {
    const { props } = this;
    const { className, children, style, prefixCls } = props;

    const cls = classnames({
      [`${prefixCls}-body-item`]: true,
      [`${prefixCls}-body-item-active`]: this.state.selected,
      [className!]: !!className,
    });

    return (
      <div className={cls} style={style}>
        {children}
      </div>
    );
  }
}

export default Tab;
