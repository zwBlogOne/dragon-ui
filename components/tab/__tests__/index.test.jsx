import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tab from '../index';

describe('Tab', () => {
  it('renders normal Tab correctly', () => {
    const wrapper = render(
      <div>
        <Tab.Group defaultValue={1}>
          <Tab title="选项卡1">
            <div style={{ padding: 10 }}>
              这是选项卡1的文字
            </div>
          </Tab>
          <Tab title="选项卡2">
            <div style={{ padding: 10 }}>
              这是选项卡2的文字
            </div>
          </Tab>
          <Tab title="选项卡3">
            <div style={{ padding: 10 }}>
              这是选项卡3的文字
            </div>
          </Tab>
        </Tab.Group>
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Tab correctly with more props', () => {
    const wrapper = render(
      <div>
        <Tab.Group defaultValue={1} isRadius theme="error">
          <Tab title="选项卡1">
            <div style={{ padding: 10 }}>
              这是选项卡1的文字
            </div>
          </Tab>
          <Tab title="选项卡2">
            <div style={{ padding: 10 }}>
              这是选项卡2的文字
            </div>
          </Tab>
          <Tab title="选项卡3">
            <div style={{ padding: 10 }}>
              这是选项卡3的文字
            </div>
          </Tab>
        </Tab.Group>
      </div>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when receiving new value', () => {
    const wrapper = mount(
      <Tab.Group defaultValue={1}>
        <Tab title="选项卡1">
          <div style={{ padding: 10 }}>
            这是选项卡1的文字
          </div>
        </Tab>
        <Tab title="选项卡2">
          <div style={{ padding: 10 }}>
            这是选项卡2的文字
          </div>
        </Tab>
        <Tab title="选项卡3">
          <div style={{ padding: 10 }}>
            这是选项卡3的文字
          </div>
        </Tab>
      </Tab.Group>
    );

    expect(wrapper.state('value')).toEqual(1);
    wrapper.setProps({ value: 2 });
    expect(wrapper.state('value')).toEqual(2);
  });

  it('behaves correctly when changing tab', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Tab.Group defaultValue={1} onChange={onChange}>
        <Tab title="选项卡1">
          <div style={{ padding: 10 }}>
            这是选项卡1的文字
          </div>
        </Tab>
        <Tab title="选项卡2">
          <div style={{ padding: 10 }}>
            这是选项卡2的文字
          </div>
        </Tab>
        <Tab title="选项卡3">
          <div style={{ padding: 10 }}>
            这是选项卡3的文字
          </div>
        </Tab>
      </Tab.Group>
    );

    wrapper.find('.ui-tab-header-item').at(2).simulate('click');
    expect(onChange).toHaveBeenCalledWith(2);
  });
});

