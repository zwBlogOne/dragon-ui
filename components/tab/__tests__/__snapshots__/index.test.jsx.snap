// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab renders Tab correctly with more props 1`] = `
<div>
  <div
    class="ui-tab radius theme-error"
  >
    <ul
      class="ui-tab-header"
    >
      <li
        class="ui-tab-header-item"
      >
        选项卡1
      </li>
      <li
        class="ui-tab-header-item active"
      >
        选项卡2
      </li>
      <li
        class="ui-tab-header-item"
      >
        选项卡3
      </li>
    </ul>
    <div
      class="ui-tab-body"
    >
      <div
        class="ui-tab-body-item"
      >
        <div
          style="padding:10px"
        >
          这是选项卡1的文字
        </div>
      </div>
      <div
        class="ui-tab-body-item ui-tab-body-item-active"
      >
        <div
          style="padding:10px"
        >
          这是选项卡2的文字
        </div>
      </div>
      <div
        class="ui-tab-body-item"
      >
        <div
          style="padding:10px"
        >
          这是选项卡3的文字
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tab renders normal Tab correctly 1`] = `
<div>
  <div
    class="ui-tab theme-default"
  >
    <ul
      class="ui-tab-header"
    >
      <li
        class="ui-tab-header-item"
      >
        选项卡1
      </li>
      <li
        class="ui-tab-header-item active"
      >
        选项卡2
      </li>
      <li
        class="ui-tab-header-item"
      >
        选项卡3
      </li>
    </ul>
    <div
      class="ui-tab-body"
    >
      <div
        class="ui-tab-body-item"
      >
        <div
          style="padding:10px"
        >
          这是选项卡1的文字
        </div>
      </div>
      <div
        class="ui-tab-body-item ui-tab-body-item-active"
      >
        <div
          style="padding:10px"
        >
          这是选项卡2的文字
        </div>
      </div>
      <div
        class="ui-tab-body-item"
      >
        <div
          style="padding:10px"
        >
          这是选项卡3的文字
        </div>
      </div>
    </div>
  </div>
</div>
`;
